### PR TITLE
test: add pulp test case

### PR DESCRIPTION
**Pull request by yih-redhat**

### CODE REVIEW

**Comment by yih-redhat** [comment]
hi @henrywang @mcattamoredhat this pr is to add pulp test case in downstream testing, ready for review now.

File: ostree-pulp.sh [hunk 1]
```diff
@@ -0,0 +1,405 @@
...
+SSH_KEY_PUB=$(cat "${SSH_KEY}".pub)
+
+# Prepare osbuild-composer repository file
+sudo mkdir -p /etc/osbuild-composer/repositories
+
+# RHEL 8.8 and CS8 is still RO for /sysroot on raw image and simplified installer
+# The RO setting on RHEL 8.8 and CS8 is not configured by ostree, but osbuild-composer
+# by PR https://github.com/osbuild/osbuild-composer/pull/3178
+case "${ID}-${VERSION_ID}" in
+    "rhel-9.3")
```
Comment (henrywang) [inline ostree-pulp.sh:37]: You can skip 9.3 and add 9.4 here.

File: ostree-pulp.sh [hunk 2]
```diff
@@ -0,0 +1,405 @@
...
+greenprint "🛃 Checking for SSH is ready to go"
+for _ in $(seq 0 30); do
+    RESULTS="$(wait_for_ssh_up $GUEST_ADDRESS)"
+    if [[ $RESULTS == 1 ]]; then
+        echo "SSH is ready now! 🥳"
+        break
+    fi
+    sleep 10
+done
+
```
Comment (henrywang) [inline ostree-pulp.sh:400]: Why no `check-ostree` playbook running in this case?

File: ostree-pulp.sh [hunk 3]
```diff
@@ -0,0 +1,405 @@
...
+greenprint "📄 ostree blueprint"
+cat "$BLUEPRINT_FILE"
+
+# Prepare the blueprint for the compose.
+greenprint "📋 Preparing blueprint"
+sudo composer-cli blueprints push "$BLUEPRINT_FILE"
+sudo composer-cli blueprints depsolve ostree
+
+# Build commit image
+build_image ostree "$IMAGE_TYPE" test "$PULP_CONFIG_FILE"
```
Comment (henrywang) [inline ostree-pulp.sh:271]: What does `test` mean? If it's image_key, you have IMAGE_KEY defined already.

**Review by henrywang — CHANGES_REQUESTED** [review]
I have some comments.

File: ostree-pulp.sh [hunk 4]
```diff
@@ -0,0 +1,405 @@
...
+# Provision the software under test.
+./setup.sh
+
+# Get OS data.
+source /etc/os-release
+ARCH=$(uname -m)
+
+# Set up variables.
+TEST_UUID=$(uuidgen)
+IMAGE_KEY="ostree-raw-${TEST_UUID}"
```
Comment (henrywang) [inline ostree-pulp.sh:13]: The ostree-raw does not make sense in this case.

File: ostree-pulp.sh [hunk 5]
```diff
@@ -0,0 +1,405 @@
...
+        OS_VARIANT="rhel9-unknown"
+        BOOT_LOCATION="http://${DOWNLOAD_NODE}/rhel-9/nightly/updates/RHEL-9/latest-RHEL-9.3.0/compose/BaseOS/x86_64/os/"
+        ;;
+    *)
+        echo "unsupported distro: ${ID}-${VERSION_ID}"
+        exit 1;;
+esac
+
+# Pulp setup
+PULP_CONFIG_FILE="$(pwd)/pulp.toml"
```
Comment (henrywang) [inline ostree-pulp.sh:49]: Please put temp file into TEMPDIR, not current folder. Or please remove it if test finished.

File: ostree-pulp.sh [hunk 6]
```diff
@@ -0,0 +1,405 @@
...
+}
+
+##################################################
+##
+## Upload ostree commit to pulp test
+##
+##################################################
+# Setup pulp server
+greenprint "📄 Setup pulp server with one container"
+mkdir -p settings pulp_storage pgsql containers
```
Comment (henrywang) [inline ostree-pulp.sh:192]: If you put those folder in current folder, please remove them after test.

File: ostree-pulp.sh [hunk 7]
```diff
@@ -0,0 +1,405 @@
...
+echo "CONTENT_ORIGIN='http://$(hostname):8080'
```
Comment (henrywang) [inline ostree-pulp.sh:193]: Why do you use hostname here? Why not IP address? Hostname can be changed for different platform.

File: ostree-pulp.sh [hunk 8]
```diff
@@ -0,0 +1,405 @@
...
+            --name pulp \
+            --volume "$(pwd)/settings":/etc/pulp:Z \
+            --volume "$(pwd)/pulp_storage":/var/lib/pulp:Z \
+            --volume "$(pwd)/pgsql":/var/lib/pgsql:Z \
+            --volume "$(pwd)/containers":/var/lib/containers:Z \
+            --device /dev/fuse \
+            quay.io/pulp/pulp:nightly
+
+# Wait until pulp service is fully functional
+sleep 120
```
Comment (henrywang) [inline ostree-pulp.sh:208]: Could you please use curl to confirm container running, not hard code wait?

File: ostree-pulp.sh [hunk 9]
```diff
@@ -0,0 +1,405 @@
...
+# Build commit image
+build_image ostree "$IMAGE_TYPE" test "$PULP_CONFIG_FILE"
+
+# Start httpd to serve ostree repo.
+greenprint "🚀 Starting httpd daemon"
+# osbuild-composer-tests have mod_ssl as a dependency. The package installs
+# an example configuration which automatically enabled httpd on port 443, but
+# that one is already in use. Remove the default configuration as it is useless
+# anyway.
+sudo rm -f /etc/httpd/conf.d/ssl.conf
```
Comment (henrywang) [inline ostree-pulp.sh:279]: Don't need this in downstream.

File: ostree-pulp.sh [hunk 10]
```diff
@@ -0,0 +1,405 @@
...
+echo "Zeroing out empty space."
+# This forces the filesystem to reclaim space from deleted files
+dd bs=1M if=/dev/zero of=/var/tmp/zeros || :
+rm -f /var/tmp/zeros
+echo "(Don't worry -- that out-of-space error was expected.)"
+
+%end
+STOPHERE
+
+sudo sed -i '/^user\|^sshkey/d' "${KS_FILE}"
```
Comment (henrywang) [inline ostree-pulp.sh:363]: If you remove user and sshkey, please remove them from ks file, not here.

File: ostree-pulp.sh (hunk 1 above)
Comment (yih-redhat) [reply ostree-pulp.sh]: fixed

File: ostree-pulp.sh (hunk 2 above)
Comment (yih-redhat) [reply ostree-pulp.sh:400]: added check-ostree test

File: ostree-pulp.sh (hunk 3 above)
Comment (yih-redhat) [reply ostree-pulp.sh:279]: the parameter test has no effect on this command, added comments inline to explain.

File: ostree-pulp.sh (hunk 4 above)
Comment (yih-redhat) [reply ostree-pulp.sh]: fixed.

File: ostree-pulp.sh (hunk 5 above)
Comment (yih-redhat) [reply ostree-pulp.sh:51]: now these files will be removed in cleanup step.

File: ostree-pulp.sh (hunk 6 above)
Comment (yih-redhat) [reply ostree-pulp.sh:198]: now these folders will be removed in cleanup step.

File: ostree-pulp.sh (hunk 7 above)
Comment (yih-redhat) [reply ostree-pulp.sh]: changed to IP address.

File: ostree-pulp.sh (hunk 8 above)
Comment (yih-redhat) [reply ostree-pulp.sh:214]: The sleep function here is not to wait container to be running, actually after the container is up and running, it still needs around 60 seconds to be full functional. 

For example, if run below command immediately after container is up and running,

podman exec -it pulp bash -c "pulpcore-manager reset-admin-password"

will get some internal errors, and wait around 60 seconds try again, it works well.

File: ostree-pulp.sh (hunk 9 above)
Comment (yih-redhat) [reply ostree-pulp.sh]: Removed

File: ostree-pulp.sh (hunk 10 above)
Comment (yih-redhat) [reply ostree-pulp.sh:363]: Removed them in ks file.

**Comment by henrywang** [comment]
@yih-redhat Did you try your new script on RHEL 9.4? Please add your manual result here. Then we can land this PR. Thanks.

**Comment by yih-redhat** [comment]
> @yih-redhat Did you try your new script on RHEL 9.4? Please add your manual result here. Then we can land this PR. Thanks.

Yes, it passed local test in my rhel9.4 machine.

**Review by henrywang — APPROVED** [review]

**Pull request merged**